### PR TITLE
Updated Palette Generation

### PR DESCRIPTION
**Pull request by trevoring-okta**
## Description:
* Odyssey 1.x expanded their palette to include 10 different hues rather than the original 5(?)
* Uses Adobe's `leonardo-contrast-colors` package (same tool Odyssey used to generate their palettes for 1.x) to generate a WCAG-friendly palette that matches Odyssey's required contrast ratios based on the `brandColors.primaryColor`config option
    * Pinned on Leonardo `1.0.0-alpha.13` because that is the last stable version. Also, Leonardo does not have proper TS support for their main API, so we had to define types via `leonardo.d.ts`. These types were taken from [a PR](https://github.com/adobe/leonardo/pull/207) that has been merged into Leonardo but for some reason not been published to NPM. Ideally this will get published soon and we can bump the package and remove the type definitions
* Also starts passing customer-defined tokens that can be passed in custom code editor via `theme.tokens` into `OdysseyProvider`'s `designTokensOverride`, otherwise token overrides are not accessible via the `useOdysseyDesignTokens` hook

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [x] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-667943](https://oktainc.atlassian.net/browse/OKTA-667943)

#### Switching from Odyssey 1.x default palette to custom green palette
https://github.com/okta/okta-signin-widget/assets/106110543/acc6edae-0c0f-41c9-bb94-eaf41ab638d4

### CODE REVIEW

File: src/v3/src/components/Images/AppIcon.tsx [hunk 1]
```diff
@@ -10,20 +10,24 @@
...
+      sx={{
+        color: Tokens.PalettePrimaryMain,
+        width: '16px',
+        height: '16px',
+      }}
+      titleAccess={loc('icon.title.application', 'login')}
+      aria-hidden
+    />
+  );
+};
```
Comment (trevoring-okta) [inline src/v3/src/components/Images/AppIcon.tsx:33]: Updated custom icons that Denys added to use token overrides instead of `theme.palette`. We're trying to remove all uses of `theme` in favor of tokens

File: src/v3/src/components/Widget/GlobalStyles.tsx [hunk 2]
```diff
@@ -13,36 +13,40 @@
...
-);
+const GlobalStyles: FunctionComponent = () => {
+  const Tokens = useOdysseyDesignTokens();
+
+  return (
+    <MuiGlobalStyles
+      styles={svgStyles(Tokens)}
+    />
+  );
+};
```
Comment (trevoring-okta) [inline src/v3/src/components/Widget/GlobalStyles.tsx:50]: See other comment about moving all uses of `theme` to use tokens when possible

File: src/v3/src/util/theme.ts [hunk 3]
```diff
@@ -10,170 +10,126 @@
...
-  if (customTokens) {
-    if (customTokens.PalettePrimaryMain) {
-      const p = generatePalette(customTokens.PalettePrimaryMain);
-      set(theme, 'palette.primary.lighter', customTokens.PalettePrimaryLighter ?? p.lighter);
-      set(theme, 'palette.primary.light', customTokens.PalettePrimaryLight ?? p.light);
-      set(theme, 'palette.primary.main', customTokens.PalettePrimaryMain ?? p.main);
-      set(theme, 'palette.primary.dark', customTokens.PalettePrimaryDark ?? p.dark);
-      set(theme, 'palette.primary.contrastText', p.contrastText);
-    }
-    if (customTokens.PaletteDangerMain) {
```
Comment (trevoring-okta) [inline src/v3/src/util/theme.ts:139]: There was some miscommunication with the pre-1.x implementation of palette generation. We thought we should be generating palettes whenever the customer passed in token overrides for any of the `Main` hues but @alexdahl-okta clarified that palette generation should only be via `brandColors.primaryColor`

File: src/v3/src/util/theme.ts [hunk 4]
```diff
@@ -10,170 +10,126 @@
...
+    tokensOverride.PalettePrimaryDark ??= p?.CustomHue700;
+    tokensOverride.PalettePrimaryDarker ??= p?.CustomHue800;
+    tokensOverride.PalettePrimaryHeading ??= p?.CustomHue900;
   }
 
+  // Odyssey 1.x does not export their theme, but we can recreate it
+  const baseOdysseyTheme = createOdysseyMuiTheme({
+    odysseyTokens: { ...Tokens, ...tokensOverride },
+  });
+
```
Comment (trevoring-okta) [inline src/v3/src/util/theme.ts:120]: One niche thing to note: palette generation is now limited to the design tokens layer and will not apply at the base MUI theming layer (the base MUI theme appears in code via `theme.palette.*`). This is fine because even though Odyssey does create a MUI theming layer via this `createOdysseyMuiTheme` call, they use their design tokens to achieve essentially all of their styling. 

TLDR: If the customer begins seeing any styling that can be traced to the base MUI theme layer, there's probably something wrong and this theming is just a fallback

File: src/v3/src/util/theme.ts [hunk 5]
```diff
@@ -10,170 +10,128 @@
...
+    tokensOverride.PalettePrimaryHeading ??= p?.CustomHue900;
   }
 
+  const mergedTokens = { ...Tokens, ...tokensOverride };
+
+  // Odyssey 1.x does not export their theme, but we can recreate it
+  const baseOdysseyTheme = createOdysseyMuiTheme({
+    odysseyTokens: mergedTokens,
+  });
+
```
Comment (trevoring-okta) [inline src/v3/src/util/theme.ts:122]: One niche thing to note: palette generation is now limited to the design tokens layer and will not apply at the base MUI theming layer (the base MUI theme appears in code via theme.palette.*). This is fine because even though Odyssey does create a MUI theming layer via this createOdysseyMuiTheme call, they use their design tokens to achieve essentially all of their styling.

TLDR: If the customer begins seeing any styling that can be traced to the base MUI theme layer, there's probably something wrong and the styling is just a fallback

File: src/v3/package.json [hunk 6]
```diff
@@ -39,15 +39,16 @@
...
     "@emotion/cache": "^11.9.3",
     "@emotion/react": "^11.9.3",
     "@emotion/styled": "^11.9.3",
     "@hcaptcha/react-hcaptcha": "^1.8.1",
     "@mui/icons-material": "^5.8.4",
     "@mui/material": "^5.8.5",
-    "@okta/odyssey-design-tokens": "^1.2.0",
-    "@okta/odyssey-react-mui": "^1.2.0",
+    "@okta/odyssey-design-tokens": "1.9.1",
+    "@okta/odyssey-react-mui": "1.9.1",
```
Comment (lesterchoi-okta) [inline src/v3/package.json:51]: You're intentionally pinning to the exact semver?

File: src/v3/screenshots/base/UI_demo/UI_demo_RTL_VRT.png
Comment (lesterchoi-okta) [inline src/v3/screenshots/base/UI_demo/UI_demo_RTL_VRT.png:1]: These VRT screenshots are really useful. Is the background color of secondary button changing due to palette generation or a separate change in Odyssey?

**Review by lesterchoi-okta — APPROVED** [review]

File: src/v3/screenshots/base/UI_demo/UI_demo_RTL_VRT.png
Comment (denysoblohin-okta) [reply src/v3/screenshots/base/UI_demo/UI_demo_RTL_VRT.png:1]: Looks like it's a change in Odyssey 1.5.0 https://github.com/okta/odyssey/blob/main/CHANGELOG.md#150-2023-10-27

File: src/v3/src/util/theme.ts (hunk 5 above)
Comment (trevoring-okta) [reply src/v3/src/util/theme.ts:122]: That scenario would probably be fixed in SIW and would be a case where the SIW dev decided to use the base MUI `theme` API rather than the Odyssey `useOdysseyDesignTokens` hook

File: src/v3/package.json (hunk 6 above)
Comment (trevoring-okta) [reply src/v3/package.json:51]: Yeah I decided to pin exactly because even though Odyssey promised no breaking changes they have still been having some breaking changes. For example, the `OdysseyProvider` API was broken for a few versions and only just got fixed at this `1.9.1` update

File: src/v3/src/util/theme.ts [hunk 7]
```diff
@@ -10,170 +10,128 @@
...
- * If WHITE vs color has a contrast ratio <4.5:1 this will return BLACK.
- * Contrast ratios are symmetrical using the WCAG 2.x algorithm
- *
- * @param color color to compare against black/white
- */
-const getInverseTextColor = (color: string): string => (
-  chroma.contrast(color, WHITE_HEX) >= 4.5 ? WHITE_HEX : BLACK_HEX
-);
+// Odyssey-defined contrast ratios for their WCAG-friendly palettes
+const ODYSSEY_RATIOS = [1.1, 1.31, 1.61, 2.22, 3.32, 4.5, 4.95, 8.72, 11.73, 14.94];
```
[the remark this answers is not in the document]
Comment (trevoring-okta) [reply src/v3/src/util/theme.ts:25]: Yeah these numbers have been specifically defined by Odyssey and design

File: src/v3/src/components/Widget/GlobalStyles.tsx [hunk 8]
```diff
@@ -13,36 +13,40 @@
...
   },
 });
 
-const GlobalStyles: FunctionComponent = () => (
-  <MuiGlobalStyles
-    styles={(theme) => svgStyles(theme)}
-  />
-);
+const GlobalStyles: FunctionComponent = () => {
+  const Tokens = useOdysseyDesignTokens();
```
[the remark this answers is not in the document]
Comment (lesterchoi-okta) [reply src/v3/src/components/Widget/GlobalStyles.tsx:43]: Nope not a blocker. Just pointing it out because TitleCase is usually reserved for classes, types, interface, and enums(?), i.e., wanted to prevent ambiguity or confusion.
